// schema-typed/lib/NumberType.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Type2 = require('./Type');

var _Type3 = _interopRequireDefault(_Type2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

function FN(value) {
  return +value;
}

var NumberType = function (_Type) {
  _inherits(NumberType, _Type);

  NumberType.from = function from(n) {
    return n;
  };

  function NumberType() {
    var errorMessage = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Please enter a valid number';

    _classCallCheck(this, NumberType);

    var _this = _possibleConstructorReturn(this, _Type.call(this, 'number'));

    _Type.prototype.pushCheck.call(_this, function (value) {
      return (/^-?(?:\d+|\d{1,3}(?:,\d{3})+)(?:\.\d+)?$/.test(value)
      );
    }, errorMessage);
    return _this;
  }

  NumberType.prototype.isInteger = function isInteger(errorMessage) {
    _Type.prototype.pushCheck.call(this, function (value) {
      return (/^-?\d+$/.test(value)
      );
    }, errorMessage);
    return this;
  };

  NumberType.prototype.pattern = function pattern(regexp, errorMessage) {
    _Type.prototype.pushCheck.call(this, function (value) {
      return regexp.test(value);
    }, errorMessage);
    return this;
  };

  NumberType.prototype.isOneOf = function isOneOf(numLst, errorMessage) {
    _Type.prototype.pushCheck.call(this, function (value) {
      return numLst.includes(FN(value));
    }, errorMessage);
    return this;
  };

  NumberType.prototype.range = function range(min, max, errorMessage) {
    _Type.prototype.pushCheck.call(this, function (value) {
      return FN(value) >= min && FN(value) <= max;
    }, errorMessage);
    return this;
  };

  NumberType.prototype.min = function min(_min, errorMessage) {
    _Type.prototype.pushCheck.call(this, function (value) {
      return FN(value) >= _min;
    }, errorMessage);
    return this;
  };

  NumberType.prototype.max = function max(_max, errorMessage) {
    _Type.prototype.pushCheck.call(this, function (value) {
      return FN(value) <= _max;
    }, errorMessage);
    return this;
  };

  return NumberType;
}(_Type3.default);

exports.default = function (errorMessage) {
  return new NumberType(errorMessage);
};
