// rsuite/lib/ControlLabel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("./utils");

var ControlLabel =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(ControlLabel, _React$Component);

  function ControlLabel() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = ControlLabel.prototype;

  _proto.render = function render() {
    var controlId = (0, _get2.default)(this.context, 'formGroup.controlId');
    var _this$props = this.props,
        _this$props$htmlFor = _this$props.htmlFor,
        htmlFor = _this$props$htmlFor === void 0 ? controlId : _this$props$htmlFor,
        srOnly = _this$props.srOnly,
        className = _this$props.className,
        classPrefix = _this$props.classPrefix,
        rest = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["htmlFor", "srOnly", "className", "classPrefix"]);
    var classes = (0, _classnames.default)(classPrefix, className, {
      'sr-only': srOnly
    });
    return React.createElement("label", (0, _extends2.default)({}, rest, {
      htmlFor: htmlFor,
      className: classes
    }));
  };

  return ControlLabel;
}(React.Component);

(0, _defineProperty2.default)(ControlLabel, "contextTypes", {
  formGroup: _propTypes.default.object
});
(0, _defineProperty2.default)(ControlLabel, "handledProps", ["className", "classPrefix", "htmlFor", "srOnly"]);

var _default = (0, _utils.defaultProps)({
  classPrefix: 'control-label'
})(ControlLabel);

exports.default = _default;
