// rsuite/lib/FlexboxGridItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("./utils");

var FlexboxGridItem =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(FlexboxGridItem, _React$Component);

  function FlexboxGridItem() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = FlexboxGridItem.prototype;

  _proto.render = function render() {
    var _classNames;

    var _this$props = this.props,
        className = _this$props.className,
        classPrefix = _this$props.classPrefix,
        colspan = _this$props.colspan,
        order = _this$props.order,
        props = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["className", "classPrefix", "colspan", "order"]);
    var addPrefix = (0, _utils.prefix)(classPrefix);
    var classes = (0, _classnames.default)(classPrefix, className, addPrefix(colspan), (_classNames = {}, _classNames[addPrefix("order-" + order)] = order, _classNames));
    return React.createElement("div", (0, _extends2.default)({}, props, {
      className: classes
    }));
  };

  return FlexboxGridItem;
}(React.Component);

(0, _defineProperty2.default)(FlexboxGridItem, "defaultProps", {
  colspan: 0,
  order: 0
});
(0, _defineProperty2.default)(FlexboxGridItem, "handledProps", ["className", "classPrefix", "colspan", "order"]);

var _default = (0, _utils.defaultProps)({
  classPrefix: 'flex-box-grid-item'
})(FlexboxGridItem);

exports.default = _default;
