// rsuite/lib/Navbar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _setStatic = _interopRequireDefault(require("recompose/setStatic"));

var _NavbarBody = _interopRequireDefault(require("./NavbarBody"));

var _NavbarHeader = _interopRequireDefault(require("./NavbarHeader"));

var _utils = require("./utils");

var Navbar =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(Navbar, _React$Component);

  function Navbar() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = Navbar.prototype;

  _proto.getChildContext = function getChildContext() {
    return {
      navbar: this.props.hasChildContext
    };
  };

  _proto.render = function render() {
    var _this$props = this.props,
        className = _this$props.className,
        Component = _this$props.componentClass,
        classPrefix = _this$props.classPrefix,
        appearance = _this$props.appearance,
        rest = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["className", "componentClass", "classPrefix", "appearance"]);
    var addPrefix = (0, _utils.prefix)(classPrefix);
    var unhandled = (0, _utils.getUnhandledProps)(Navbar, rest);
    var classes = (0, _classnames.default)(classPrefix, addPrefix(appearance), className);
    return React.createElement(Component, (0, _extends2.default)({}, unhandled, {
      className: classes,
      role: "navigation"
    }));
  };

  return Navbar;
}(React.Component);

(0, _defineProperty2.default)(Navbar, "defaultProps", {
  hasChildContext: true,
  appearance: 'default'
});
(0, _defineProperty2.default)(Navbar, "childContextTypes", {
  navbar: _propTypes.default.bool
});
(0, _defineProperty2.default)(Navbar, "handledProps", ["appearance", "className", "classPrefix", "componentClass", "hasChildContext"]);
var EnhancedNavbar = (0, _utils.defaultProps)({
  componentClass: 'div',
  classPrefix: 'navbar'
})(Navbar);
(0, _setStatic.default)('Header', _NavbarHeader.default)(EnhancedNavbar);
(0, _setStatic.default)('Body', _NavbarBody.default)(EnhancedNavbar);
var _default = EnhancedNavbar;
exports.default = _default;
