// rsuite/lib/Notification.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _curry2 = _interopRequireDefault(require("lodash/curry"));

var React = _interopRequireWildcard(require("react"));

var _rsuiteNotification = require("rsuite-notification");

var _Icon = _interopRequireDefault(require("./Icon"));

var _constants = require("./utils/constants");

var _prefix = require("./utils/prefix");

var classPrefix = (0, _prefix.getClassNamePrefix)() + "notification";
var defaultOptions = {
  classPrefix: classPrefix
};

_rsuiteNotification.Notify.config(defaultOptions);

function appendIcon(type, content) {
  if (!_constants.STATUS_ICON_NAMES[type]) {
    return content;
  }

  return React.createElement("div", {
    className: classPrefix + "-title-with-icon"
  }, React.createElement(_Icon.default, {
    icon: _constants.STATUS_ICON_NAMES[type]
  }), content);
}

appendIcon.handledProps = [];

function proxy(type, config) {
  config.title = appendIcon(type, config.title);

  _rsuiteNotification.Notify[type](config);
}

var sendMessage = (0, _curry2.default)(proxy);
var _default = {
  open: sendMessage('open'),
  info: sendMessage('info'),
  success: sendMessage('success'),
  warning: sendMessage('warning'),
  error: sendMessage('error'),
  remove: function remove(key) {
    _rsuiteNotification.Notify.remove(key);
  }
};
exports.default = _default;
