// dom-lib/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  animation: true,
  transition: true,
  translateDOMPositionXY: true,
  getVendorPrefixedName: true,
  BrowserSupportCore: true,
  DOMMouseMoveTracker: true,
  WheelHandler: true
};
Object.defineProperty(exports, "transition", {
  enumerable: true,
  get: function get() {
    return _transition2.default;
  }
});
Object.defineProperty(exports, "translateDOMPositionXY", {
  enumerable: true,
  get: function get() {
    return _translateDOMPositionXY2.default;
  }
});
Object.defineProperty(exports, "getVendorPrefixedName", {
  enumerable: true,
  get: function get() {
    return _getVendorPrefixedName2.default;
  }
});
Object.defineProperty(exports, "BrowserSupportCore", {
  enumerable: true,
  get: function get() {
    return _BrowserSupportCore2.default;
  }
});
Object.defineProperty(exports, "DOMMouseMoveTracker", {
  enumerable: true,
  get: function get() {
    return _DOMMouseMoveTracker2.default;
  }
});
Object.defineProperty(exports, "WheelHandler", {
  enumerable: true,
  get: function get() {
    return _WheelHandler2.default;
  }
});
exports.animation = void 0;

var _classNames = require("./classNames");

Object.keys(_classNames).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _classNames[key];
    }
  });
});

var _style = require("./style");

Object.keys(_style).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _style[key];
    }
  });
});

var _query = require("./query");

Object.keys(_query).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _query[key];
    }
  });
});

var _events = require("./events");

Object.keys(_events).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _events[key];
    }
  });
});

var _animation = _interopRequireWildcard(require("./animation"));

exports.animation = _animation;

var _transition2 = _interopRequireDefault(require("./transition"));

var _translateDOMPositionXY2 = _interopRequireDefault(require("./transition/translateDOMPositionXY"));

var _getVendorPrefixedName2 = _interopRequireDefault(require("./getVendorPrefixedName"));

var _BrowserSupportCore2 = _interopRequireDefault(require("./BrowserSupportCore"));

var _DOMMouseMoveTracker2 = _interopRequireDefault(require("./DOMMouseMoveTracker"));

var _WheelHandler2 = _interopRequireDefault(require("./WheelHandler"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }
