// material-ui-slider/es5/libs/markdown/canvas.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _construct2 = _interopRequireDefault(require("@babel/runtime/helpers/construct"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _marked = _interopRequireDefault(require("marked"));

var _standalone = require("@babel/standalone");

var _editor = _interopRequireDefault(require("../editor"));

var Canvas =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(Canvas, _React$Component);

  function Canvas(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Canvas);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Canvas).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "playerId", "".concat(parseInt(Math.random() * 1e9, 10).toString(36)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "description", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "extras", {});
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "source", []);
    _this.state = {
      showBlock: false
    };
    var matchs = props.children.match(/([^]*)[\r\n|\n](```[^]+```)/);

    if (matchs && matchs[1]) {
      var descs = matchs[1].match(/([^]*)[\r\n|\n]===\s?extra\s?[\r\n|\n]([^]+)[\r\n|\n]===/);
      if (!descs) _this.description = (0, _marked.default)(matchs[1]);else {
        _this.description = (0, _marked.default)(descs[1]);
        descs[2].split(/\r\n/).forEach(function (desc) {
          var value = desc.split(/\s+/);
          _this.extras[value[0]] = value[1];
        });
      }
    }

    if (matchs && matchs[2]) {
      _this.source = matchs[2].match(/```(.*)[\n|\r\n]([^]+)[\r\n|\n]```/);
    }

    return _this;
  }

  (0, _createClass2.default)(Canvas, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.renderSource(this.source[2]);
    }
  }, {
    key: "blockControl",
    value: function blockControl() {
      this.setState({
        showBlock: !this.state.showBlock
      });
    }
  }, {
    key: "renderSource",
    value: function renderSource(value) {
      var _this2 = this;

      import('../../src').then(function (Element) {
        var args = ['context', 'React', 'ReactDOM'];
        var argv = [_this2, _react.default, _reactDom.default];

        for (var key in Element) {
          args.push(key);
          argv.push(Element[key]);
        }

        return {
          args: args,
          argv: argv
        };
      }).then(function (_ref) {
        var args = _ref.args,
            argv = _ref.argv;
        var code = (0, _standalone.transform)("\n        class Demo extends React.Component {\n          ".concat(value, "\n        }\n\n        ReactDOM.render(<Demo {...context.props} />, document.getElementById('").concat(_this2.playerId, "'))\n      "), {
          presets: ['es2015', 'react']
        }).code;
        args.push(code);
        var Fn = Function;
        (0, _construct2.default)(Fn, (0, _toConsumableArray2.default)(args)).apply(_this2, argv);
        _this2.source[2] = value;
      }).catch(function (err) {
        if (process.env.NODE_ENV !== 'production') {
          throw err;
        }
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      return _react.default.createElement("div", {
        className: "demo-block demo-box demo-".concat(this.props.name)
      }, _react.default.createElement("div", {
        className: "demo-block-control"
      }, this.extras.codepen && _react.default.createElement("a", {
        className: "control-button",
        href: this.extras.codepen,
        target: "_black"
      }, _react.default.createElement("svg", {
        "aria-labelledby": "simpleicons-codepen-icon",
        role: "img",
        viewBox: "0 0 24 24"
      }, _react.default.createElement("title", {
        id: "simpleicons-codepen-icon"
      }, "CodePen icon"), _react.default.createElement("path", {
        d: "M24 8.182l-.018-.087-.017-.05c-.01-.024-.018-.05-.03-.075-.003-.018-.015-.034-.02-.05l-.035-.067-.03-.05-.044-.06-.046-.045-.06-.045-.046-.03-.06-.044-.044-.04-.015-.02L12.58.19c-.347-.232-.796-.232-1.142 0L.453 7.502l-.015.015-.044.035-.06.05-.038.04-.05.056-.037.045-.05.06c-.02.017-.03.03-.03.046l-.05.06-.02.06c-.02.01-.02.04-.03.07l-.01.05C0 8.12 0 8.15 0 8.18v7.497c0 .044.003.09.01.135l.01.046c.005.03.01.06.02.086l.015.05c.01.027.016.053.027.075l.022.05c0 .01.015.04.03.06l.03.04c.015.01.03.04.045.06l.03.04.04.04c.01.013.01.03.03.03l.06.042.04.03.01.014 10.97 7.33c.164.12.375.163.57.163s.39-.06.57-.18l10.99-7.28.014-.01.046-.037.06-.043.048-.036.052-.058.033-.045.04-.06.03-.05.03-.07.016-.052.03-.077.015-.045.03-.08v-7.5c0-.05 0-.095-.016-.14l-.014-.045.044.003zm-11.99 6.28l-3.65-2.44 3.65-2.442 3.65 2.44-3.65 2.44zm-1.034-6.674l-4.473 2.99L2.89 8.362l8.086-5.39V7.79zm-6.33 4.233l-2.582 1.73V10.3l2.582 1.726zm1.857 1.25l4.473 2.99v4.82L2.89 15.69l3.618-2.417v-.004zm6.537 2.99l4.474-2.98 3.613 2.42-8.087 5.39v-4.82zm6.33-4.23l2.583-1.72v3.456l-2.583-1.73zm-1.855-1.24L13.042 7.8V2.97l8.085 5.39-3.612 2.415v.003z"
      }))), this.extras.github && _react.default.createElement("a", {
        className: "control-button",
        href: this.extras.github,
        target: "_black"
      }, _react.default.createElement("svg", {
        "aria-labelledby": "simpleicons-github-icon",
        role: "img",
        viewBox: "0 0 24 24"
      }, _react.default.createElement("title", {
        id: "simpleicons-github-icon"
      }, "GitHub icon"), _react.default.createElement("path", {
        d: "M12 .297c-6.63 0-12 5.373-12 12 0 5.303 3.438 9.8 8.205 11.385.6.113.82-.258.82-.577 0-.285-.01-1.04-.015-2.04-3.338.724-4.042-1.61-4.042-1.61C4.422 18.07 3.633 17.7 3.633 17.7c-1.087-.744.084-.729.084-.729 1.205.084 1.838 1.236 1.838 1.236 1.07 1.835 2.809 1.305 3.495.998.108-.776.417-1.305.76-1.605-2.665-.3-5.466-1.332-5.466-5.93 0-1.31.465-2.38 1.235-3.22-.135-.303-.54-1.523.105-3.176 0 0 1.005-.322 3.3 1.23.96-.267 1.98-.399 3-.405 1.02.006 2.04.138 3 .405 2.28-1.552 3.285-1.23 3.285-1.23.645 1.653.24 2.873.12 3.176.765.84 1.23 1.91 1.23 3.22 0 4.61-2.805 5.625-5.475 5.92.42.36.81 1.096.81 2.22 0 1.606-.015 2.896-.015 3.286 0 .315.21.69.825.57C20.565 22.092 24 17.592 24 12.297c0-6.627-5.373-12-12-12"
      }))), _react.default.createElement("button", {
        className: "control-button",
        onClick: this.blockControl.bind(this)
      }, _react.default.createElement("svg", {
        "aria-labelledby": "simpleicons-coding-icon",
        role: "img",
        viewBox: "0 0 24 24"
      }, _react.default.createElement("g", null, _react.default.createElement("path", {
        d: "M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
      }))))), this.state.showBlock && _react.default.createElement("div", {
        className: "meta"
      }, this.description && _react.default.createElement("div", {
        ref: "description",
        className: "description",
        dangerouslySetInnerHTML: {
          __html: this.description
        }
      }), _react.default.createElement(_editor.default, {
        value: this.source[2],
        theme: this.props.theme,
        lineNumbers: true,
        onChange: function onChange(code) {
          return _this3.renderSource(code);
        }
      })), _react.default.createElement("div", {
        className: "source",
        id: this.playerId
      }));
    }
  }]);
  return Canvas;
}(_react.default.Component);

var _default = Canvas;
exports.default = _default;
