// rsuite/lib/DateRangePicker/Calendar/Calendar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("../../utils");

var _MonthDropdown = _interopRequireDefault(require("../../Calendar/MonthDropdown"));

var _Header = _interopRequireDefault(require("../../Calendar/Header"));

var _View = _interopRequireDefault(require("./View"));

var _is_after = _interopRequireDefault(require("date-fns/is_after"));

var _add_months = _interopRequireDefault(require("date-fns/add_months"));

var Calendar =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(Calendar, _React$Component);

  function Calendar() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMoveForword", function () {
      var onMoveForword = _this.props.onMoveForword;
      onMoveForword && onMoveForword((0, _add_months.default)(_this.getPageDate(), 1));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMoveBackward", function () {
      var onMoveBackward = _this.props.onMoveBackward;
      onMoveBackward && onMoveBackward((0, _add_months.default)(_this.getPageDate(), -1));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "disabledBackward", function () {
      var _this$props = _this.props,
          calendarDate = _this$props.calendarDate,
          index = _this$props.index;
      var after = (0, _is_after.default)(calendarDate[1], (0, _add_months.default)(calendarDate[0], 1));

      if (index === 0) {
        return false;
      }

      if (!after) {
        return true;
      }

      return false;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "disabledForword", function () {
      var _this$props2 = _this.props,
          calendarDate = _this$props2.calendarDate,
          index = _this$props2.index;
      var after = (0, _is_after.default)(calendarDate[1], (0, _add_months.default)(calendarDate[0], 1));

      if (index === 1) {
        return false;
      }

      if (!after) {
        return true;
      }

      return false;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "disabledMonth", function (date) {
      var _this$props3 = _this.props,
          calendarDate = _this$props3.calendarDate,
          value = _this$props3.value,
          index = _this$props3.index,
          disabledDate = _this$props3.disabledDate;
      var after = true;

      if (disabledDate && disabledDate(date, value, 'MONTH')) {
        return true;
      }

      if (index === 1) {
        after = (0, _is_after.default)(date, calendarDate[0]);
        return !after;
      }

      after = (0, _is_after.default)(calendarDate[1], date);
      return !after;
    });
    return _this;
  }

  var _proto = Calendar.prototype;

  _proto.getPageDate = function getPageDate() {
    var _this$props4 = this.props,
        calendarDate = _this$props4.calendarDate,
        index = _this$props4.index;
    return calendarDate[index];
  };

  _proto.shouldMountDate = function shouldMountDate(props) {
    var _ref = props || this.props,
        format = _ref.format;

    return /Y/.test(format) && /M/.test(format) && /D/.test(format);
  };

  _proto.render = function render() {
    var _classNames;

    var _this$props5 = this.props,
        calendarState = _this$props5.calendarState,
        onSelect = _this$props5.onSelect,
        onMouseMove = _this$props5.onMouseMove,
        onToggleMonthDropdown = _this$props5.onToggleMonthDropdown,
        onChangePageDate = _this$props5.onChangePageDate,
        disabledDate = _this$props5.disabledDate,
        className = _this$props5.className,
        value = _this$props5.value,
        hoverValue = _this$props5.hoverValue,
        isoWeek = _this$props5.isoWeek,
        limitEndYear = _this$props5.limitEndYear,
        classPrefix = _this$props5.classPrefix,
        rest = (0, _objectWithoutPropertiesLoose2.default)(_this$props5, ["calendarState", "onSelect", "onMouseMove", "onToggleMonthDropdown", "onChangePageDate", "disabledDate", "className", "value", "hoverValue", "isoWeek", "limitEndYear", "classPrefix"]);
    var pageDate = this.getPageDate();
    var dropMonth = calendarState === 'DROP_MONTH';
    var addPrefix = (0, _utils.prefix)(classPrefix);
    var calendarClasses = (0, _classnames.default)(classPrefix, className, (_classNames = {}, _classNames[addPrefix('show-month-dropdown')] = dropMonth, _classNames));
    var unhandled = (0, _utils.getUnhandledProps)(Calendar, rest);
    return React.createElement("div", (0, _extends2.default)({}, unhandled, {
      className: calendarClasses
    }), React.createElement(_Header.default, {
      showMonth: true,
      date: pageDate,
      disabledBackward: this.disabledBackward(),
      disabledForword: this.disabledForword(),
      onMoveForword: this.handleMoveForword,
      onMoveBackward: this.handleMoveBackward,
      onToggleMonthDropdown: onToggleMonthDropdown
    }), React.createElement(_View.default, {
      activeDate: pageDate,
      value: value,
      hoverValue: hoverValue,
      onSelect: onSelect,
      onMouseMove: onMouseMove,
      disabledDate: disabledDate,
      isoWeek: isoWeek
    }), React.createElement(_MonthDropdown.default, {
      date: pageDate,
      show: dropMonth,
      disabledMonth: this.disabledMonth,
      onSelect: onChangePageDate,
      limitEndYear: limitEndYear
    }));
  };

  return Calendar;
}(React.Component);

(0, _defineProperty2.default)(Calendar, "defaultProps", {
  calendarDate: [new Date(), (0, _add_months.default)(new Date(), 1)],
  index: 0
});
(0, _defineProperty2.default)(Calendar, "handledProps", ["calendarDate", "calendarState", "className", "classPrefix", "disabledDate", "format", "hoverValue", "index", "isoWeek", "limitEndYear", "onChangePageDate", "onMouseMove", "onMoveBackward", "onMoveForword", "onSelect", "onToggleMonthDropdown", "value"]);
var enhance = (0, _utils.defaultProps)({
  classPrefix: 'calendar'
});

var _default = enhance(Calendar);

exports.default = _default;
